### ISSUE TITLE
BL redesign

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#7]
**Reported by @cristianbogdan on 7 Oct 2002 13:25 UTC**
the BEST Business Logic (BL) starts to be messy. we need to take it as an
example and try to redesign the makumba BL support and lookup in a way that
allows for better-organized BLs.

the BL should be more connected to the MDD (data model) and less connected to
the JSP layer (view). Fred has proposed sth called "business objects". mb his
mail should be attached here

there is a tendency in the current BEST BL to have one BL file per operation,
mostly due to authentication reasons. more operations should be grouped in one
file, so probably separate authentication should be allowed per operation
(rather than per BL file)

another issue here is the number of DB accesses made. organizing the BL methods
nicely (i.e. without repeating code) can result in too many queries to the db.
maybe OQL _fragments_ should be worked with, instead of OQL _accesses_

Migrated-From: http://trac.makumba.org/ticket/7

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 7 Oct 2002 13:26 UTC**
setting the target milestone to designDecission as the implementation is not
clear, hence more discussion is needed

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 17 Dec 2002 00:04 UTC**
i would summarize on of the objective here as follows: look in karamba and find
all <% scriplets%> that have to do with data processing, as well as find all
workarounds in the BL, try to generalize them and support them in the BL

many of the scriplets that have to do with data presentation will have to be
generalized and supported as part of a new JSP taglib (bug 5)

would be nice for the scriplet+workaround-hunting + generalization work to start
ASAP

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 May 2004 12:22 UTC**
below comes a mail i sent recently.
i would like to do something about this before 0.6 . will see if it's possible

while coming back to makumba on a fresh mind i realised that the current BL
design is a mixture between controller (processing data entered by the user) and
model (processing data in the database according to business rules).

one of the main problems is that the same database type (e.g.
best.johnny.Student) can be processed with more than one BL code assigned from
different parts of the site. for example i can always write a
test/deleteJohnnyAccount.jsp and make a BL for it without knowing that there is
already such an operation done in another BL class, and probably i'll forget
something that's already taken care of there, etc. this is primarily due to the
fact that we associate the BL more with the JSP page (controller) than with the
MDD file (model).

of course both the controller and the model parts are needed. every form needs a
bit of preparation of its data before sending it to the model, and a bit of
processing of the results from the model (especially errors) before displaying
results to the user.

i think more separation between the controller and the model is in order. also,
they should be easy to find by the makumba programmer.

so here are my suggestions (note that authentication is not yet settled in the
design below):

1) write the controller part in the form page itself. can be written as a  <%!
java class-level scriplet %> which can be associated with a form in a page based
on the form name. the controller can have methods to pre-process data from the
form and to process messages at response time (e.g. the current message= can be
made more sophisticated, etc).

for example:
<mak:form name="doSomething" .... >
<%!
public void doSomethingPreProces(Dictionary data, Database db){
  // add or remove sth from data, maybe use stuff in the db
}
public String doSomethingShowMessage(Object result){
   // process the result returned by the BL in case of success and return some HTML
}
public String doSomethingShowException(Exception e){
   // process the exception returned by the BL and return some HTML
}
%>

2) write the model code (business rules) in the MDD file, or in a java file with
the same name as the MDD file, in the same directory. alternatively, as proposed
in bug 8, we might support the writing of BL in XML (still, since learning Java
is of importance for developers, it should always be available, and the first to
be supported)

public void onNew(Dictionary data, Database db, Pointer auth){
   // note that there are no more "attributes" as the pre-processing should have
been done by now.
   // not sure who/how should indicate the authentication, probably the
controller part.
}

3) other ideas:
- look at jakarta Struts before finishing this design
- maybe try to design authentication/authorization so that the standard servlet
context authentication is used instead of makumba-specific
- try to design the whole thing so that it can co-exist with present-style
business logic
- if a clean design can be achieved and implemented, it would be nice to have it
done fast, so the re-visiting/refactoring of karamba code can be done after.
- look at bugs 7 and 8, maybe even 9 before going on with design.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 May 2004 12:24 UTC**
(From update of attachment 7)
>Date: Wed, 18 Sep 2002 22:56:17 +0200
>From: Frederik Habils <frederik.habils@gmx.net>
>To: makumba@BEST.eu.org
>Subject: [MAKUMBA] application design: business object logic ?

>hi folks,
>
>
>i've been thinking a bit, about what I believe is a general issue in
>makumba apps.
>
>I'll just present "in short" my ideas... maybe they spark some interest,
>maybe somebody will say "oh, no, don't do that", etc :-)
>
>
>In short, the fact seems to me, 
>- that we are benefitting from OQL in that we can store objects in the db. 
>- but we are not making real objects out of the MDD-types.
>- thereby making life more difficult, by spreading object logic around in
>tons of places.
>
>
>Example:
>-------------------
>It struck me how often, in johnny jsp-pages, i have to recreate the
>vivaldi season logic, out of the pure data: "get season confirmation,
>compare with today, check if event is inSeason, if not get application DL
>of event, calculate confirmation DL as application DL + 21 days for
>non-optimised out-of-ssn event"; "check if event ended and student
>confirmed", etc, etc. 
>
>Instead of writing the exact code over and over again, for whatever query
>on the application, activity, (business objects) ... I would rather have
>some "object logic" to call:
>	season.isPastXxxDl()
>	act.isEnded()
>	act.followsSeason()
>	application.isExtraAccepted()
>	application.hasConfirmed()
>
>etc.
>
>So, I started on an Activity object
>http://18080.makumba.org/fred-k/classes/org.eu.best.johnny.Activity
>
>.. it mostly provides logic around the season and event timeline.
>
>
>Thoughts
>-------------------
>
>I'm not saying that there is sth missing in Makumba... it's just that I
>hadn't looked at this till a couple of days ago, and I felt sth missing in
>my approach of JohnnyTS programming. There's a lot of focus in Makumba on
>the Rapid Application Development (get data in jsp, make a few
>calculations on them there if necessary, display) which is just great. 
>
>Makumba stresses also Business Logic which I would imagine as sort of a
>"horizontal" logic (a bit of logic connected to every page => spreading
>around object logic in small pieces across the BL), while Business Objects
>I would see as "vertical" logic (they cut through all layers of the
>business logic, providing consistent object logic) 
>
>I believe such objects, while not necessary for every MDD-type, can be
>useful for those business objects that are frequently used and behind
>which data a lot of logic is hidden.
>
>
>
>>From there, I'm currently STUCK on following:
>--------------------------------------------
>
>- using business Objects likely means more accesses to the database (e.g.
>Activity constructor does 2 (very simple, no joins) db accesses.). E.g. if
>I don't use the Activity object in a johnny page about an activity, there
>is 1 access to the db from the jsp, that collects all the info in one
>query. If I have an additional Activity object, to help simplify the logic
>coding, then I have additional queries to construct the Activity object
>(queries that actually retrieve the same data from db, as the mak:object
>already did in the jsp)
>
>- So, while I'm not sure if these extra db-accesses are significant, I was
>wondering (also regarding performance), if it would be good to try and
>keep those objects alive across page accesses, e.g. by making them
>session-attributes. That would mean that, e.g. a student browsing his
>johnny pages (who can see a max of 6 activities in one (welcome page), if
>there are 2 seasons overlapping), would not reconstruct these Activity
>objects all the time, when moving to other pages, but would keep them
>during a session. 
>
>Fact is that I don't know too much about SessionScope variables (where are
>they stored, when are they deleted, are there memory considerations?,
>..), so I won't look at this yet. 
>
>- Finally, I'm wondering whether to make the objects as light as possible
>(providing only methods for the most intensively used logic around the
>object, and keeping only that part of the data in the MDD-type, that is
>necessary for those methods, OR to provide all possible methods, and to
>store all the fields of the MDD-type in the object). At the moment, I did
>the Activity object as a 'lightweight' object, so e.g. not storing
>Activity#name, #lbg, #description, etc... but storing #start, #end, etc.
>
>- And perhaps, they could even be used to encapsulate a number of SQL
>queries that are executed on these object, in BL... ? 
>
>
>
>well, that's it :-)
>
>curiously awaiting your comments ;-)
>
>Fred.
>
>
>
>
>

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 May 2004 12:28 UTC**
(From update of attachment 8)
>Date: Thu, 19 Sep 2002 10:52:56 +0200
>From: Cristian Bogdan <cristi@nada.kth.se>
>To: makumba@BEST.eu.org
>Subject: Re: [MAKUMBA] application design: business object logic ?

>Hi Fred
>
>i am glad you started to think of this. johnny98 was built around such 
>business objects but they got mixed with a cache, then pressure from 
>users amounted and no refactoring was made as the application was too 
>cumbersome to sart up, hence too difficult to develop on, plus it had a 
>cumbersome way to store data (java serialized objects) that obliged us 
>to have a cache, etc etc
>
>we also thought of these things when designing makumba but we didn't 
>know how to get them right (there is a biig makumba.html in the doc, 
>it's full of visionary ideas). the problem when doing design is that you 
>can't grasp the precise needs of the user/setting. also there, we had 
>application (TS) pressure mounting, so we did "the simplest thing that 
>could possibly work" (see xprogramming.com) and decided to "grow it out 
>of needs".
>
>and here you are, expressing the needs :)
>
>when growing sth out of needs, you always need to re-write your apps, 
>moving things from the apps to the base technology. so from the outset 
>we must ask: is ITD prepared to do rewrites? (see also JSP 2) i have 
>promoted the 'continuous rewrite' idea as a natural way of doing things. 
>instead of re-doing the entire app seldomly (like we did with johnny) 
>it's much better to re-do _often_ parts of the app (refactoring) like we 
>think of the BL now. so such a change should not modify any JSP!
>
>when we thought of business objects before we couldn't go too far since 
>we didn't know the application code, as it didn't exist :). now it 
>exists and you know it, so the design problem is much easier.
>
>Model 1 (short term)
>- it may be that some tweaking in the BL discovery code, some 
>restructuring of the current BL classes (grouping methods differently in 
>new classes, taking type name away from method names and having that 
>type name as the name of the class), and a re-mapping in 
>MakumbaController.prop can achieve a business-object like structure, 
>where each form response method is a method in a BO class.
>- so maybe what should be worked on is simply the convention for 
>discovering the method called by a form. what changes is that the BL 
>classes are closer in structure to MDD types than they are to JSP pages 
>(i.e. one BO per MDD rather than 1-N BLs per JSP dir), which is very good.
>- BO methods should not map to one query each. that is too low 
>granularity. performance is critically dependent on the number of queries
>- so the BO methods should e.g. return strings that represent parts of 
>OQL queries instead of actually doing queries and returning data. this 
>would reduce the number of queries by postponing query execution until 
>more of its elements are knwon.
>- a method to asemble such "query components" could be part of makumba. 
>it is possible to model OQL fragments (query components) in Java and 
>provide a method query(QueryComponent f) where f is a composition of 
>many small OQL fragments. that would spare us from OQL compiler work 
>(needed below), but i'm not sure if sth else would be achieved. 
>manipulating OFs can be quite cumbersome for the mak db API user
>
>
>Model 2 (longer term)
>- Model 2 is based on model 1 but eliminates java and moves everything 
>in the MDD
>- the road from MDD data to java data (record.get("fieldname")) and back 
>to MDD (record.put("fieldname") can be eliminated. java is too far from 
>the MDD, and the Map transition that we make is not very natural. if we 
>want to improve the transition (e.g. each db field is a java field or 
>method), we'll end up re-inventing very complex commercial products that 
>already exist, plus the jakarta OJB or JDO (which is sun but looks quite 
>badly promoted??), which implement http://ODMG.org in java.
>- i basically suggest to stick to our simplicity, and implement parts of 
>ODMG in MDD + OQL + maybe a small procedural lang. OQL is already from ODMG
>- simple methods should exist in or around the MDD. MDD forces us to 
>keep them simple
>- ideally they should be called like act.isEnded() in an OQL query and 
>the OQL manager should be able to expand them as needed
>- ideally methods should be writable in the MDD as OQL expressions so as 
>not to introduce a new language. however small procedural parts are needed.
>- one question is, should logic be separate from the data (to e.g. not 
>disturb feature designers) or together with it (more OO). i think the OO 
>thing is better, and the 'disturbance' can be negotiated with a bit of 
>nice editting and commenting
>
>i also am more and more convinced that, whatever their representation 
>(MDD or Java) business objects should never contain any data, just data 
>manipulation methods (e.g. stuff that changes db data) and query 
>component manipulation techniques. they should therefore be 
>_business_object_strategies_ (BOS). BL objects today are the same (they 
>are strategies), and they work pretty well (they avoid caches). once 
>data is represented in both the db and in an object in Java, you run 
>into the trouble of caching (see below).
>
>i have already struggled with some of the ideas above in my "XML BL" 
>mail some time ago. find it attached. in the BO approach, instead of writing
>	<operation op="on_new" type="best.johnny.Application"
>		name="app">
>you can write in b.j.Application just
>	<operation op="on_new" >
>and assume that the name of the object is... "this" :)
>
>as you see the resulting language is much shorter because the work of 
>going from the DB to java and back is circumvented.
>	
>
>short comments to some of the stuff will folow below. before that, what 
>can be done next? i think the way to go is
>- don't reinvent the wheel. look at ODMG, OJB, JDO and databases that 
>support objects methods (e.g. i think oracle). see what we can get from 
>there. or maybe we can decide that we just switch to one of them...
>- if you see fit, continue to develop your current approach and see 
>where it leads
>- think of how BL/BO/BOS/XML code would look like for BEST's BL layer in 
>model1, model2 or some other alternative. see what comes more natural! 
>shorter and simple to write. it's cool to expriment with languages that 
>don't exist yet :)) send scriplets to the list (see my xml bl message)
>- take into account more than johnny, think of other BLs
>
>Frederik Habils wrote:
>> - that we are benefitting from OQL in that we can store objects in the db. 
>you mean, from MDD?
>
>
>> the Rapid Application Development (get data in jsp, make a few
>> calculations on them there if necessary, display) which is just great. 
>
>yep, i think it helped get an app written fast. but it doesn't allow for 
>much refactoring, getting repeated code in one place (Xprogramming 
>principle "once and only once", etc)
>
>
>> Makumba stresses also Business Logic which I would imagine as sort of a
>> "horizontal" logic (a bit of logic connected to every page => spreading
>
>actually the BL as i thought of it originally was supposed not to be 
>one-logic-per-page-or-per-JSPdir but more like a business object thing 
>(BO). but when i gave it to developers, they thought of it natural like 
>that 'parallel' with the JSP structure. i am not accusing anybody here 
>except the designer :), but this is not healthy, the model should not 
>know anything about views, let alone being written in the same structure 
>as the view.
>
>> - using business Objects likely means more accesses to the database (e.g.
>
>i suggested how this can be reduced
>
>> - So, while I'm not sure if these extra db-accesses are significant, I was
>they are :)
>
>> wondering (also regarding performance), if it would be good to try and
>> keep those objects alive across page accesses, e.g. by making them
>> session-attributes. That would mean that, e.g. a student browsing his
>
>hmmmmm. if the objects are just strategy objects (i.e. they don't store 
>data) that's ok. if they do store data, it's problematic, because then 
>they are a cache, and caches must be built and maintained consistent 
>with the original data. and then you go into the whole johnny cache 
>problems we had before
>
>while i do agree that caches will be of use, i suggest to go one step at 
>a time and leave caches for a while
>
>> Fact is that I don't know too much about SessionScope variables (where are
>> they stored, when are they deleted, are there memory considerations?,
>> ...), so I won't look at this yet. 
>
>they are stored by the server in a way compliant with the server-neutral 
>servlet API. they are also stored on disk as serializable java objects 
>(a binary format which we used to store johnny data before), if the 
>server dies, they are loaded from disk at restart
>
>> store all the fields of the MDD-type in the object). At the moment, I did
>> the Activity object as a 'lightweight' object, so e.g. not storing
>> Activity#name, #lbg, #description, etc... but storing #start, #end, etc.
>
>i think you are struggling with mixing the concept of cache with the 
>concept of business object, hence you need make these tradeoffs.
>
>> - And perhaps, they could even be used to encapsulate a number of SQL
>> queries that are executed on these object, in BL... ? 
>
>i really think they should encapsulate query fragments that can be used 
>in queries when needed
>
>cristi
>
>Date: Mon, 04 Feb 2002 13:52:54 +0100
>From: Cristian Bogdan <cristi@nada.kth.se>
>To: makumba@best.eu.org
>Subject: * idea: XML business logic
>
>Hi
>
>This may be a longer term idea, but I think it's worth thinking of it.
>The point is to define the business logic in XML instead of Java,
>staying Java-neutral at all levels (except JSP where Java is just a
>scripting language). 
>
>This message will illustrate how this can be done for the two business
>logic functions: authentication and db operations accompanying code
>(data validation, etc)
>
>1) Authentication. Say we miss a secret attribute called
>"johnnyStudent". 
>
>Instead of writing a Java 
>Object findJohnnyStudent(Database db, Attributes a)...
>
>one could write
>
><find attr="johnnyStudent" type="best.johnny.Student">
>    <find attr="auth" type="best.Auth" >
>	<where>
>	
>	  auth.username=$username AND auth.password=$password 
>		OR
>
>	  $checkCodes($username, $password) AND 
>	  EXISTS (SELECT c FROM auth.codes c WHERE c.code= $username 
>			AND c.season=$season)
>        </where>
>	<notFoundError>
>		bad code pair or bad username/password
>	</notFoundError>
>	
>
>    </find>
>    <where>
>	johnnyStudent.student.authentication= auth
>    </where>
>    <notFoundError>you do not have a johnny account</notFoundError>
>    
></find>
>
>This is based on "query components" same as the JSP pages (the OQL  uses
>subselects, but it's not hard to get subselects to work in the current
>OQL translator). In the
>example above, everything can be executed in one query.
>
>2) Data validation by a business logic method can be easy. Instead
>of on_newBestJohnnyApplication(), one can write:
>
><operation op="on_new" type="best.johnny.Application" name="app">
>     <check error="you have already applied to this activity">
>	NOT EXISTS(SELECT app2 FROM best.johnny.Application app2 WHERE 
>		$app.activity=app2.activity)
>     </check>
>
>     
>
>     <check error="you can only apply to 3 activities in a season">
>	COUNT (SELECT app2 FROM best.johnny.Application app2 
>	WHERE $app.activity.season=app2.activity.season) >3
>     </check>
>
>     
>     <check error="the application deadline has passed">
>	(SELECT deadline FROM $app.activity.season.deadlines WHERE
>deadline.name="application") < $currentDate()
>     </check>
></operation>
>
>At the implementation level, all checks are boolean OQL SELECT
>projections, so all needed checks can be combined in one query, and that
>query can actually be the same with the "find" (authentication) query
>above... and everything can often be combined with the first query of
>the JSP page :)
>
>One small problem here is synchronization (2 simmultaneous checks of
>number of applications when the respective student applied to 2
>activities, and in general, operations taking place at the same time),
>can be fixed (at SQL client level or maybe even in the SQL engine)
>something like:
>
><operation op="new" type="best.johnny.Application" name="app"
>   lockOn="$johnnyStudent" >
>
>i.e. only execute this operation one-at-a-time for a given value of
>$johnnyStudent
>
>3) For non-elementary operations (insert/change more records than the
>one indicated, or responding to generic forms that don't have a main
>record), then things get a bit more complicated, but not impossible:
>
>- need for a scripting language (e.g. for getting the curent date, for
>checking a pair of johnny codes, for complicated generic forms like
>johnny ranking or list adjustment, for the famous johnny optimization). 
>
>Ant supports "BSF" languages for scripting (see the <DEFANGED_script> target of
>ant, and http://oss.software.ibm.com/developerworks/opensource/bsf/ for
>BSF languages, one of them is Javascript). 
>
>- need for database insert/update/delete, which can be easily defined as
>XML tags, something like
>
><insert type="best.johnny.Application.xxx" name="yyy">
>	<set>yyy.xxx=aaa.bbb</set> (set as in "assign", I know it's confusing)
>..
></insert>
>
><find from="T1 x, T2 y">
><where>condition</where>
><set>xxx.yyy=zzz.bbb</set> ... (or <delete/> )
></find>
>(the notation above can be easily implemented with JDBC2 updatable
>result sets)
>
>4) design discussion:
>
>OK, so in this case the BL will become
>- a set of XML files, maybe keeping the same BL-discovery mechanism we
>have now
>- eliminates need for Java knowledge
>- gets rid of the Java notations for makumba data & attributes
>(Dictionary, Vector, Enumeration, Attributes) making the code smaller,
>while not loosing strict type checking at OQL level
>- could co-exist with Java business logic, so a Java business logic such
>as best's could be gradually (file-by-file) written in the new form
>(gradually because it's not hard to write the XML BL, it's harder to
>make the XML BL engine work like the Java one used to :)
>- easy to change at server runtime (maybe too easy :)
>- since queries can be combined, it can generate a minimum number of
>queries to the SQL engine, resulting to the best possible performance in
>terms of SQL and Java-SQL communication
>- will probably be slower in the scripting part (the scripts are
>interpreted by Java)
>- but gives the high-performance possibility to generate SQL procedures
>from the script (by hacking the BSF interpretor) and compile them in the
>SQL engine.
>- also, as far as best needs go, scripts would be rarely used (ranking,
>optimization, list adjustment are an infime percentage of the accesses)
>
>tell me what u think
>
>cristi
>

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 3 May 2004 12:28 UTC**
(From update of attachment 15)
>Date: Fri, 05 Jul 2002 11:21:23 +0200
>From: Cristian Bogdan <cristi@nada.kth.se>
>To: frederik.habils@BEST.eu.org
>Cc: makumba@BEST.eu.org
>
>Hi Fred
>
>(this has not much to do with helpdesk as with naming conventions)
>
>i remember that i designed the current BL lookup convention thinking of 
>having one BL per dir to the max... if one works like that, the name of 
>the individual JSP file doesn't matter as far as BL lookup is concerned
>
>however, in practice it seems that some JSPs sometimes need different 
>access rights than the rest of the dir (therefore they need a different 
>BL). even so, i still think that, either the dirs should be organized 
>some other way, either the BL convention needs be changed to reduce the 
>number of Java files, and keep related BL code together in a class 
>rather than disparated in several classes.
>
>basically the conclusion is that the current conventions (for both BL 
>lookup and JSP naming) need be revised, but this should be done 
>_based_on_the_experience_ which we didn't have when we first made them. 
>that is, we should revise these conventions by looking at the existing 
>page names and BLs
>
>cristi
>
>Frederik Habils wrote:
>
>>So you would have to copy the general logic about Surveys in
>>ViewSurveyLogic and EditSurveyLogic, while if the names would be
>>	/SurveyView
>>	/SurveyEdit
>>then you can combine certain logic in the SurveyLogic handler, and the
>>special logic in SurverEditLogic and SurveyViewLogic.
>>  
>>
>this is a very good example.   if you'd use a survey directory you'd 
> have survey/view.jsp, survey/edit.jsp, which will lead to the same BL name.
>
>>TECH: I've understood (?) : Makumba will find only ONE handler per page
>>(the one with longest name). But typically XxxYyyLogic will extend
>>(inherit) XxxLogic, thus inheriting all it's methods. So in _layman's_
>>terms, you could say that 'all possible' handler classes are found, if
>>existing.
>>  
>>
>yes, if the respective BLs are inherited, then you could say that all 
>are used.
>
>cristi
>
>
>
>
>
>

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 23 Dec 2007 17:55 UTC**
more and more recent thoughts on the topic:

-------- Original Message --------
Subject: [business logic redesign
Date: Thu, 08 Nov 2007 14:12:35 +0100
From: Cristian Bogdan <cristi@nada.kth.se>
To: makumba-devel@lists.sourceforge.net <makumba-devel@lists.sourceforge.net>

1) per-object (not per page/dir) BL: while working on the new hibernate 
Transaction we realized that it's very easy to call per-object 
on_insert, on_update kind of BL. the question is, should e.g. the 
on_insert parameter of the "main" object contain also the sub-objects?

1a) the trouble with the sub-objects is that they are not visible by the 
  validation/authorization code that we envision in the MDD, so this 
looks a bit inconsistent (remember my question about 'cascading'). 
should we make them visible there somehow?

2) should we support hibernate in the current per page/dir BL? it seems 
to be easy to do since hibernate is seen now as a simple Transaction.

3) whether we support hibernate just per object or also per page/dir, it 
seems that we would gain much more if we pass the object, not a 
dictionary. in other words

on_insert(Transaction t, best.Student s)

where s has all the stuff prepared (all its related objects set), only 
it's not attached to the hib session, and getters and setters can be 
called for it. this would make the bl much more idio(ma)tic

3a) the problem is a bit different for update

on_update(Transaction t, best.Student s, Dictionary fieldsToChange)

here s would be the _present_ value
though even there, fieldsToChange can contain Objects (say 
best.internal.Student) in its values, not Pointers

the other possibility we discussed was to do the update first (so s 
contains the changes), and revert the transaction if we find a problem 
with fieldsTochange. doing the update in the db (transaction, not 
committed yet) lets us perform the validation checks against the db

----

in short, there should be a new BL soon. let's design it nicely.

--------
the other thing to discuss is how to indicate in the page or "somewhere" 
that a certain form is going to be submitted to hibernate, not makumba 
BL + db code

one way would be to use a taglib URI like www.makumba.org/forms-hibernate

still, sometimes we will want that entire directories or sets of related 
pages or entire sets of types, or all the application are treated with 
hibernate... sounds a bit illogical to link that to the taglib used. 
though the taglib works well to ensure co-existence of mak and hibernate 
pages... you can even use <jsp:include /> to include a hql page in an 
oql one or the other way around  :) 


rudi's reply:


-------- Original Message --------
Subject: Re: [Makumba-devel](Makumba-devel]) business logic redesign
Date: Sat, 17 Nov 2007 02:28:07 +0100
From: Rudolf Mayer <rudolf.mayer@student.tugraz.at>
Reply-To: rudolf.mayer@student.tugraz.at
To: Cristian Bogdan <cristi@nada.kth.se>,	"makumba-devel@lists.sourceforge.net" <makumba-devel@lists.sourceforge.net>
References: <47330B43.9010205@nada.kth.se>

hey!

Cristian Bogdan wrote:
> 1) per-object (not per page/dir) BL: while working on the new hibernate 
> Transaction we realized that it's very easy to call per-object 
> on_insert, on_update kind of BL. the question is, should e.g. the 


a quick question - where there specific reasons to not do this from the 
very beginning? did the per file/path logic just seem better then?

> on_insert parameter of the "main" object contain also the sub-objects?


i guess this depends clearly on how expensive it is to get these 
sub-objects. potentially, these could be many, though, if you have huge 
sets.

therefore, maybe limit it to ptrs only, and/or provide an easy way on 
how to load them on demand, like "object.loadSubobjects()" or so?

> 1a) the trouble with the sub-objects is that they are not visible by the 
>   validation/authorization code that we envision in the MDD, so this 
> looks a bit inconsistent (remember my question about 'cascading'). 
> should we make them visible there somehow?


well, there you can make simply an OQL fragment, right? as this is not 
that easy in the BL, we might need something equivalently easy (thus my 
suggestion above)

> 2) should we support hibernate in the current per page/dir BL? it seems 
> to be easy to do since hibernate is seen now as a simple Transaction.


how do you get the hibernate transaction? now in the handler methods, 
you get the makumba db transaction, or that depends on the config?

> 3) whether we support hibernate just per object or also per page/dir, it 
> seems that we would gain much more if we pass the object, not a 
> dictionary. in other words
> 
> on_insert(Transaction t, best.Student s)
> 
> where s has all the stuff prepared (all its related objects set), only 
> it's not attached to the hib session, and getters and setters can be 
> called for it. this would make the bl much more idio(ma)tic


that would be cool. but, would that be limited to hibernate?
also, what about extra form data that might be needed (given as hidden 
input). for that, we would still need the attributes, right?

> 3a) the problem is a bit different for update
> 
> on_update(Transaction t, best.Student s, Dictionary fieldsToChange)
> 
> here s would be the _present_ value
> though even there, fieldsToChange can contain Objects (say 
> best.internal.Student) in its values, not Pointers
> 
> the other possibility we discussed was to do the update first (so s 
> contains the changes), and revert the transaction if we find a problem 
> with fieldsTochange. doing the update in the db (transaction, not 
> committed yet) lets us perform the validation checks against the db


what about providing an on_update AND an after_update that would have 
the old / new values?

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 27 Jun 2010 19:37 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 09:16 UTC**
